module/VuFind/src/VuFind/Search/History.php

```diff
@@ -0,0 +1,88 @@
+<?php
+
+namespace VuFind\Search;
+
+use minSO;
+
+class History
+{
+
+    /**
+     * @var \VuFind\Db\Table\Search
+     */
+    protected $searchTable;
+
+    /**
+     * @var \Zend\Session\SessionManager
+     */
+    protected $sessionManager;
+
+    /**
+     * @var \VuFind\Search\Results\PluginManager
+     */
+    protected $resultsManager;
+
+    /**
+     * @var \VuFind\Search\Memory
+     */
+    protected $searchMemory;
+
+    /**
+     * History constructor
+     * @param \VuFind\Db\Table\Search $searchTable
+     * @param \Zend\Session\SessionManager $sessionManager
+     * @param \VuFind\Search\Results\PluginManager $resultsManager
+     * @param \VuFind\Search\Memory $searchMemory
+     */
+    public function __construct($searchTable, $sessionManager, $resultsManager, $searchMemory)
+    {
+        $this->searchTable = $searchTable;
+        $this->sessionManager = $sessionManager;
+        $this->resultsManager = $resultsManager;
+        $this->searchMemory = $searchMemory;
+    }
+
+    /**
+     * @param int $userId
+     * @return array
+     */
+    public function getSearchHistory($userId = null, $purged = false)
+    {
+        // Retrieve search history
+        $searchHistory = $this->searchTable->getSearches(
+            $this->sessionManager->getId(),
+            $userId
+        );
+
+        // Build arrays of history entries
+        $saved = $unsaved = [];
+
+        // Loop through the history
+        /** @var \VuFind\Db\Row\Search $current */
+        foreach ($searchHistory as $current) {
+
+            /** @var minSO $minSO */
+            $minSO = $current->getSearchObject();
+
+            // Saved searches
+            if ($current->saved == 1) {
+                $saved[] = $minSO->deminify($this->resultsManager);
+            } else {
+                // All the others...
+
+                // If this was a purge request we don't need this
+                if ($purged) {
```

Comment:
The purge search history functionality is broken by this refactoring. The `$this->params()->fromQuery('purge') == 'true'` used in the controller to check for the purge parameter is removed. I think this should instead be used to pass a value to this function, and perhaps the $purged variable should be renamed to something more clear.

It might also be more clean to separate the display and purge actions to separate functions -- that is, rather than purging as part of the retrieval of search history, instead have logic like:

<pre>
if (we need to purge) {
    purge the unsaved history
}
return the history
</pre>

That would require a little bit of refactoring, but would probably also be more efficient -- causing a single DELETE query followed by a single SELECT query, rather than one SELECT followed by potentially many DELETEs.

(Note that if you want to test this, you just need to click the "Purge unsaved searches" link on the search history page).